src/Microsoft.AspNet.Server.Kestrel/Http/Connection.cs

```diff
@@ -63,13 +65,53 @@ public Connection(ListenerContext context, UvStreamHandle socket) : base(context
             ConnectionControl = this;
         }
 
+        private void ExtractIP()
+        {
+            // storage for binary IP addresses
+            Libuv.sockaddr peer;
+            Libuv.sockaddr local;
+            // length of sockaddr structs
+            var peerLength = Marshal.SizeOf(typeof(Libuv.sockaddr));
+            var localLength = peerLength;
+            // buffers for addresses
+            var peerString = new StringBuilder(50);
+            var localString = new StringBuilder(50);
+            // storage for errors during decoding processes
+            Exception peerException;
+            Exception localException;
+
+            // get local and peer IPs in binary form
+            _socket.Libuv.tcp_getsockname((UvTcpHandle)_socket, out local, ref localLength);
+            _socket.Libuv.tcp_getpeername((UvTcpHandle)_socket, out peer, ref peerLength);
+            // decode as IPv4 addresses
+            _socket.Libuv.ip4_name(ref local, localString, 50, out localException);
```

Comment:
It's possible (and preferable) to go directly from a sockaddr to an IPAddress, you don't have to convert to the intermediate string format and then re-parse it first. For example (though the format may not be exactly the same here)
https://github.com/Microsoft/referencesource/blob/c697a4b9782dc8c85c02344a847cb68163702aa7/System/net/System/Net/SocketAddress.cs#L195